<a href="https://colab.research.google.com/github/Nasjr/Blood-Bank-System/blob/master/DataScience_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from PIL import Image
from sklearn.model_selection import train_test_split
from google.colab.patches import cv2_imshow 
from random import shuffle
import random
from sklearn.preprocessing import StandardScaler

In [2]:
from zipfile import ZipFile

dataset = '/content/drive/MyDrive/DataScience_Project/fulldata_resized (1).zip'

with ZipFile(dataset, 'r') as zip:
  zip.extractall()
  print('The dataset is extracted')

The dataset is extracted


In [3]:
dataset_path='/content/fulldata_resized'
file_names=os.listdir(dataset_path)
#print the first three file names
print(file_names[0:3])
#The data is shuffeled which is good but we are goining to make sure it is random
for i in range(5):
  random.seed(4)
  random.shuffle(file_names)
print('shffuled list: ',file_names[0:10])

['dog874.jpeg', 'chicken1468.jpeg', 'horse10.jpeg']
shffuled list:  ['chicken747.jpeg', 'dog1557.jpeg', 'chicken1306.jpeg', 'butterfly762.jpeg', 'horse1238.jpeg', 'butterfly343.jpeg', 'cat1352.jpeg', 'dog763.jpeg', 'cow232.jpeg', 'cow115.jpeg']


# Categories

**In this Dataset we have several Categories and we are going to show them below :**



In [4]:
D_counts={'dog':0,'cat':0,"butterfly":0,"horse":0,'cow':0,'chicken':0,"error":0}
for i in range(len(file_names)):
    if file_names[i][0:3]=='dog':
        D_counts['dog']+=1
    elif file_names[i][0:3]=='cat':
        D_counts['cat']+=1
    elif file_names[i][0:9]=='butterfly':
        D_counts['butterfly']+=1
    elif file_names[i][0:5]=='horse':
        D_counts['horse']+=1
    elif file_names[i][0:3]=='cow':
        D_counts['cow']+=1
    elif file_names[i][0:7]=='chicken':
        D_counts['chicken']+=1
    else:
        D_counts['error']+=1
print(D_counts)

{'dog': 1755, 'cat': 1668, 'butterfly': 1496, 'horse': 1686, 'cow': 1566, 'chicken': 1547, 'error': 0}


**As We see we have :**


*   Dogs -> 1775 image
*   cats -> 1668 image
*   butterfly -> 1496 image
*   horse -> 1686 image
*   cow -> 1566 image
*   chicken -> 1547 image



**Each Category Must be mapped to a number from (1 to n) where n is the number of categories :**


*   Dogs --> 0
*   cats --> 1
*   butterfly --> 2
*   horse --> 3
*   cow --> 4
*   chicken --> 5








In [5]:
#We need to list one for the filename and one for the labels

labels=[]
imgnames=[]

for i in range(len(file_names)):
  
  #Store the image names
  imgnames.append(file_names[i])
  
  #Store the labels of the images
  if file_names[i][0:3]=='dog':
        labels.append(0)
  elif file_names[i][0:3]=='cat':
        labels.append(1)
  elif file_names[i][0:9]=='butterfly':
        labels.append(2)
  elif file_names[i][0:5]=='horse':
        labels.append(3)
  elif file_names[i][0:3]=='cow':
        labels.append(4)
  elif file_names[i][0:7]=='chicken':
        labels.append(5)
  else:
        print('This is an error')

print(imgnames[0:7])
print(labels[0:7])

['chicken747.jpeg', 'dog1557.jpeg', 'chicken1306.jpeg', 'butterfly762.jpeg', 'horse1238.jpeg', 'butterfly343.jpeg', 'cat1352.jpeg']
[5, 0, 5, 2, 3, 2, 1]


In [6]:
import cv2
import glob

image_directory = '/content/fulldata_resized/'


files = []

for name in file_names:
  files.append(image_directory+name)
#check for file names
print(files[0:3])


#convert all images into a MD numpy array aka 'Tensor'
all_images_array=np.asarray([cv2.imread(file) for file in files] )


['/content/fulldata_resized/chicken747.jpeg', '/content/fulldata_resized/dog1557.jpeg', '/content/fulldata_resized/chicken1306.jpeg']


In [7]:
print(all_images_array.shape)

(9718, 244, 244, 3)


In [8]:
#shape the data into x and y variables for training and testing

X=all_images_array
y=np.asarray(labels)


# It Is time to split our data into Training and testing Data for our model using (train_test_split) function :

In [9]:
x_train,x_test,y_train,y_test=train_test_split(X,y, test_size=0.33,stratify=y,random_state=42)

In [10]:
print("training images: ")
print(x_train.shape,y_train.shape)
print('-'*50)
print("testing images: ")
print(x_test.shape,y_test.shape)


training images: 
(6511, 244, 244, 3) (6511,)
--------------------------------------------------
testing images: 
(3207, 244, 244, 3) (3207,)


In [11]:
#take the first 4000 train image and scale them
x_train_scaled_4000img=x_train[0:4000]/255

In [12]:
#take the first 3000 test image and scale them
x_test_scaled_2000img=x_test[0:2000]/255

**Scalling the Data values so that the range is between 0-1 to avoid overfitting and avoid giving weights imbalanced importance values :**

# Model Creation : 